AK/Vector.h

```diff
@@ -143,17 +145,17 @@ requires(!IsRvalueReference<T>) class Vector {
             return data()[i];
     }
 
-    ALWAYS_INLINE T const& operator[](size_t i) const { return at(i); }
-    ALWAYS_INLINE T& operator[](size_t i) { return at(i); }
+    ALWAYS_INLINE VisibleType const& operator[](size_t i) const { return at(i); }
+    ALWAYS_INLINE VisibleType& operator[](size_t i) { return at(i); }
 
-    T const& first() const { return at(0); }
-    T& first() { return at(0); }
+    VisibleType const& first() const { return at(0); }
+    VisibleType& first() { return at(0); }
 
-    T const& last() const { return at(size() - 1); }
-    T& last() { return at(size() - 1); }
+    VisibleType const& last() const { return at(size() - 1); }
+    VisibleType& last() { return at(size() - 1); }
 
     template<typename TUnaryPredicate>
-    Optional<T> first_matching(TUnaryPredicate predicate)
+    Optional<VisibleType> first_matching(TUnaryPredicate predicate)
```

Comment:
This (and anything that returns `Optional<VisibleType>`) will copy the value referred to by the element, which is _definitely_ not the expected behaviour.
Ideally this would hand you an `Optional<Foo&>`, but that's not supported.

I'd suggest either leaving it as `T` and letting the compiler complain when this is used, or disabling these functions on `Vector<Foo&>` while `Optional<Foo&>` doesn't make sense (just a simple `requires(!contains_reference)`)

Reply:
Right, I will put the required clause in there.